### PR TITLE
distributed XR-Transformer fine-tuning

### PR DESCRIPTION
**Pull request by jiong-zhang**
*Issue #, if available:* NA

*Description of changes:* 
* Add implementation of distributed XR-Transformer fine-tuning with deepspeed.
* Refactor XMC training data construction, add `XMCTextTensorizer` and `XMCLabelTensorizer` that support both pre-generation or batch-time generation of text tensors and label tensors.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pecos/distributed/diagnostic_util/deepspeed_test.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        "--verbose-level",
+        type=int,
+        choices=logging_util.log_levels.keys(),
+        default=2,
+        metavar="INT",
+        help=f"the verbose level, {', '.join([str(k) + ' for ' + logging.getLevelName(v) for k, v in logging_util.log_levels.items()])}. Default 2",
+    )
+
+    parser.add_argument(
+        "--local_rank",
```
Comment (hallogameboy) [inline pecos/distributed/diagnostic_util/deepspeed_test.py:51]: Not sure if it would be better to explain more details on terms about distributed computing. Ditto to similar arguments in other changes and terminologies like world.

File: pecos/distributed/diagnostic_util/deepspeed_test.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        type=int,
+        default=0,
+        metavar="RANK",
+        help="distributed rank",
```
Comment (hallogameboy) [inline pecos/distributed/diagnostic_util/deepspeed_test.py]: Not sure if it would be better to explain more concepts on terms in the help msg to users who are not familiar with distributed computing. Ditto to similar arguments in other changes and terminologies like world.

File: pecos/distributed/xmc/xtransformer/model.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        train_params=None,
+        pred_params=None,
+        dist_params=None,
+        **kwargs,
+    ):
+        """Train the XR-Transformer model with the given input data.
+
+        Args:
+            prob (MLProblemWithText): ML problem to solve.
+            clustering (ClusterChain, optional): preliminary hierarchical label tree,
```
Comment (hallogameboy) [inline pecos/distributed/xmc/xtransformer/model.py:134]: Some pydoc comments follow 80-char limit, but others don't. We probably can make them consistent.

File: pecos/distributed/xmc/xtransformer/model.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            dist_params = cls.DistParams.from_dict(dist_params)
+
+        if not train_params.only_encoder:
+            LOGGER.warning(
+                f"Distributed fine-tuning is only for encoder, fall back to only_encoder=True"
+            )
+            train_params.only_encoder = True
+
+        if not train_params.do_fine_tune:
+            LOGGER.warning(f"Why do you call me if you don't want fine-tune? Do fine-tune anyway..")
```
Comment (hallogameboy) [inline pecos/distributed/xmc/xtransformer/model.py]: We can also indicate the corresponding parameters as the warning message of only_encoder like "Fall back to do_fine_tune=True", so the user can exactly understand the issue in their configuration.

File: pecos/distributed/xmc/xtransformer/model.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+                parent_model.text_encoder.config.hidden_size, num_labels
+            )
+            parent_model.text_encoder.config.num_labels = num_labels
+        return parent_model
+
+    @classmethod
+    def train(
+        cls,
+        prob,
+        clustering=None,
```
Comment (hallogameboy) [inline pecos/distributed/xmc/xtransformer/model.py:124]: Is the clustering required for distributed XR-Transformer?

File: pecos/distributed/xmc/xtransformer/tune_one.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
+#!/usr/bin/env python3 -u
```
Comment (hallogameboy) [inline pecos/distributed/xmc/xtransformer/tune_one.py]: Per the previous discussion, the file name "tune_one" could be ambiguous.

File: pecos/distributed/xmc/xtransformer/tune_one.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    global_step = 0
+    tr_loss, logging_loss = 0.0, 0.0
+    total_train_time, logging_elapsed = 0.0, 0.0
+    logging_steps = train_params.logging_steps
+
+    for epoch in range(1, int(train_params.num_train_epochs) + 1):
+        shard_access_order = list(range(num_shards))
+        random.shuffle(shard_access_order)
+
+        if args.shard_scheme == "ordered":
```
Comment (hallogameboy) [inline pecos/distributed/xmc/xtransformer/tune_one.py]: Shall we assert if the shard_scheme is not in ["ordered", "synchronized"]? The case would access all shards without using the distributed sampler.

File: pecos/distributed/xmc/xtransformer/tune_one.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+            )
+            LOGGER.debug(
+                f"Rank{global_rank}/{world_size}: Training data({shard_id}/{num_shards}) loaded, num_batches={len(train_dataloader)}"
+            )
+
+            for batch_cnt, batch in enumerate(train_dataloader):
+                start_time = time.time()
+                batch = tuple(t.to(ds_engine.device) for t in batch)
+                inputs = {
+                    "input_ids": batch[0],
```
Comment (hallogameboy) [inline pecos/distributed/xmc/xtransformer/tune_one.py]: I wonder if using dict in the data format for DataLoader instead of accessing by indices. Could future inputs be an issue?

File: pecos/distributed/diagnostic_util/deepspeed_test.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3 -u
```
Comment (hallogameboy) [inline pecos/distributed/diagnostic_util/deepspeed_test.py]: Per the previous discussion, the file name and located directory can be modified with other utility functions related to distributed computing.

File: pecos/xmc/xtransformer/matcher.py [hunk 10]
```diff
@@ -950,6 +975,54 @@ def concat_features(X_feat, X_emb, normalize_emb=True):
             raise TypeError(f"Expected CSR or ndarray, got {type(X_feat)}")
         return X_cat
 
+    @staticmethod
+    def prepare_data(prob, label_pad=None, pre_tensorize_labels=False, input_transform=None):
+        """Prepare data for transformer encoder from given problem
+
+        Args:
+            prob (MLProblemWithText): problem to build data for
+            label_pad (int, optional): padding index in label embeddings.
```
Comment (hallogameboy) [inline pecos/xmc/xtransformer/matcher.py]: Would "label_padding_idx" or just "padding_idx" (following the manner of PyTorch API*) or "idx_padding" (following our manner in XMCLabelTensorizer)be better?

* https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html#torch.nn.Embedding

**Review by hallogameboy — COMMENTED** [review]
Had a pass for the initial review of the PR for distributed XR-Transformer fine-tuning.

File: pecos/distributed/xmc/xtransformer/tune_one.py (hunk 7 above)
Comment (jiong-zhang) [reply pecos/distributed/xmc/xtransformer/tune_one.py]: If shard_scheme is not in ["ordered", "synchronized"], the argparser would raise error.

File: pecos/distributed/xmc/xtransformer/model.py (hunk 5 above)
Comment (jiong-zhang) [reply pecos/distributed/xmc/xtransformer/model.py:124]: Add the logic to fall back to OVA training if clustering is not given.

File: pecos/distributed/xmc/xtransformer/model.py (hunk 4 above)
Comment (jiong-zhang) [reply pecos/distributed/xmc/xtransformer/model.py]: Updated

File: pecos/distributed/xmc/xtransformer/model.py (hunk 3 above)
Comment (jiong-zhang) [reply pecos/distributed/xmc/xtransformer/model.py:134]: Fixed

File: pecos/distributed/diagnostic_util/deepspeed_test.py (hunk 2 above)
Comment (jiong-zhang) [reply pecos/distributed/diagnostic_util/deepspeed_test.py]: Fixed

File: pecos/distributed/xmc/xtransformer/tune_one.py (hunk 6 above)
Comment (jiong-zhang) [reply pecos/distributed/xmc/xtransformer/tune_one.py]: Changed to `dist_trainer`

File: pecos/distributed/diagnostic_util/deepspeed_test.py (hunk 9 above)
Comment (jiong-zhang) [reply pecos/distributed/diagnostic_util/deepspeed_test.py]: changed to `diagnostic_tools/deepspeed_comm.py`

**Review by hallogameboy — APPROVED** [review]
